src/Compiler/Utilities/illib.fs

```diff
@@ -648,6 +648,19 @@ module List =
         | Some x -> x :: l
         | _ -> l
 
+    let rec vMapFold<'T, 'State, 'Result> (mapping: 'State -> 'T -> struct('Result * 'State)) state list : struct('Result list * 'State) =
+        match list with
+        | [] -> [], state
+        | [h] ->
+            let struct(x', s') = mapping state h
+            [x'], s'
+        | h :: t ->
+            let struct(mappedHead, stateHead) = mapping state h
+            let struct(mappedTail,stateTail) = vMapFold mapping stateHead t
```

Comment:
This isn't tail recursive, it's probably worth making it so?

Reply:
Good idea. Comes at a slight cost of having to reverse the result at the end, but it is better to have it tail recursive.